### PR TITLE
RavenDB-21523 - zstd compression for snapshot backups

### PR DESCRIPTION
**Pull request by ppekrol**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21523

### Type of change

- Optimization
- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.

### CODE REVIEW

**Comment by ppekrol** [comment]
25.6GB Database folder:
1. Zstd Optimal => 1:41 => 1.98 GB
2. Deflate Optimal => 3:50 => 2.99 GB
3. Zstd Fastest => 0:56 => 2.18 GB
4. Deflate Fastest => 3:36 => 3.31 GB
5. NoCompression => 1:11 => 25.5 GB

**Comment by ppekrol** [comment]
Restore:

1. Deflate : 0:45
2. Zstd: 0:32

File: src/Raven.Server/Config/Categories/BackupConfiguration.cs [hunk 1]
```diff
@@ -64,10 +65,10 @@ public class BackupConfiguration : ConfigurationCategory
         [ConfigurationEntry("Backup.Azure.Legacy", ConfigurationEntryScope.ServerWideOnly)]
         public bool AzureLegacy { get; set; }
 
-        [Description("Compression algorithm that is used to perform backups (does not apply to snapshot backups).")]
-        [DefaultValue(ExportCompressionAlgorithm.Zstd)]
+        [Description("Compression algorithm that is used to perform backups.")]
+        [DefaultValue(BackupCompressionAlgorithm.Zstd)]
         [ConfigurationEntry("Backup.Compression.Algorithm", ConfigurationEntryScope.ServerWideOrPerDatabase)]
-        public ExportCompressionAlgorithm CompressionAlgorithm { get; set; }
+        public BackupCompressionAlgorithm CompressionAlgorithm { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/BackupConfiguration.cs:72]: this config is for 'Smuggler' backups, name enum 'BackupCompressionAlgorithm'
introduce a new option for Snapshot backups, 'Backup.Snapshot.Compression.Algorithm' with values:
- Zstd
- Deflate

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 2]
```diff
@@ -1207,14 +1209,14 @@ private IEnumerable<FullBackup.StorageEnvironmentInformation> GetAllStoragesForB
             {
                 using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
                 {
-                    var zipArchiveEntry = package.CreateEntry(RestoreSettings.SmugglerValuesFileName, compressionLevel);
+                    var zipArchiveEntry = package.CreateEntry(RestoreSettings.SmugglerValuesFileName, CompressionLevel.NoCompression);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: switch ZipArchive to BackupZipArchive

File: src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs [hunk 3]
```diff
@@ -611,9 +611,10 @@ private InternalBackupResult CreateLocalBackupOrSnapshot(PeriodicBackupStatus st
 
                         var totalSw = Stopwatch.StartNew();
                         var sw = Stopwatch.StartNew();
+                        var compressionAlgorithm = _configuration.SnapshotSettings?.CompressionAlgorithm ?? _database.Configuration.Backup.CompressionAlgorithm;
                         var compressionLevel = _configuration.SnapshotSettings?.CompressionLevel ?? CompressionLevel.Optimal;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs]: introduce here a config option for default compression level for snapshot backups, check if we need to introduce option for smuggler backups compression level as well

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: done

File: src/Raven.Server/Config/Categories/BackupConfiguration.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Config/Categories/BackupConfiguration.cs:72]: done

File: src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs]: done

**Review by arekpalinski — APPROVED** [review]

File: src/Sparrow/Utils/ZstdStream.cs [hunk 4]
```diff
@@ -196,56 +198,113 @@ public override void Flush()
...
+                await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
             }
+
+            await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
         }
 
-        private void DisposeInternal()
+        public override async ValueTask DisposeAsync()
         {
+            if (_disposed)
```
Comment (karmeli87) [inline src/Sparrow/Utils/ZstdStream.cs:242]: I would expect that `_disposerLock` will throw if we double dispose... 
BTW, why do we have here a double dispose? is that expected?

**Review by ml054 — APPROVED** [review]

File: src/Raven.Client/Documents/Operations/Backups/SnapshotSettings.cs [hunk 5]
```diff
@@ -1,10 +1,13 @@
 using System.IO.Compression;
+using Sparrow.Backups;
 using Sparrow.Json.Parsing;
 
 namespace Raven.Client.Documents.Operations.Backups
 {
     public class SnapshotSettings : IDynamicJson
     {
+        public SnapshotBackupCompressionAlgorithm CompressionAlgorithm { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Backups/SnapshotSettings.cs]: what about backward compatibility? needs to be nullable?

Comment (grisha-kotler) [reply src/Raven.Client/Documents/Operations/Backups/SnapshotSettings.cs]: done

**Comment by ppekrol** [comment]
Can you check `SlowTests.Tests.TestsInheritanceTests.AllTestsShouldUseRavenFactOrRavenTheoryAttributes` ? Beside that LGTM

**Pull request merged**